In [1]:
import numpy as np
import sympy as sp
import scipy

# Questão 1

In [2]:
m = 1
L = 1
Ic = m*L**2
k1 = k2 = 1
M = np.array([[2*m, -m],
              [-m, m]])
K = np.array([[k1, 0],
              [0, k2]])
wns2, B = scipy.linalg.eigh(K, M)
print("wns = ")
print(np.sqrt(wns2))
phi = (1+np.sqrt(5))/2
print("phi = (1+sqrt(5))/2 = ", phi)
print("wns = ")
print("[1/phi, phi]")
print("")
print("V1 = ")
print("[phi, -1] = ", [phi, -1])
print("V2 = ")
print("[1, phi] = ", [1, phi])
print("")
print("V = ")
print("[phi,  1]")
print("[-1, phi]")
print("")
# phi = sp.symbols("phi")
V = np.array([[phi,  1],
              [-1, phi]])
ms = [3+4*phi, 3-phi]
ks = [2+phi, 2+phi]
print("ms = [3+4*phi, 3-phi]")
print("ks = [2+phi, 2+phi]")
print("B simbolico = ")
print("[phi/sqrt(3+4*phi)      1/sqrt(3-phi)]")
print("[ -1/sqrt(3+4*phi)    phi/sqrt(3-phi)]")
Bsymb = np.array([[phi/np.sqrt(3+4*phi), 1/np.sqrt(3-phi)],
                  [-1/np.sqrt(3+4*phi), phi/np.sqrt(3-phi)]])
print("B simbolico = ")
print(Bsymb)
print("B scipy = ")
print(B)

wns = 
[0.61803399 1.61803399]
phi = (1+sqrt(5))/2 =  1.618033988749895
wns = 
[1/phi, phi]

V1 = 
[phi, -1] =  [1.618033988749895, -1]
V2 = 
[1, phi] =  [1, 1.618033988749895]

V = 
[phi,  1]
[-1, phi]

ms = [3+4*phi, 3-phi]
ks = [2+phi, 2+phi]
B simbolico = 
[phi/sqrt(3+4*phi)      1/sqrt(3-phi)]
[ -1/sqrt(3+4*phi)    phi/sqrt(3-phi)]
B simbolico = 
[[ 0.52573111  0.85065081]
 [-0.3249197   1.37638192]]
B scipy = 
[[-0.52573111  0.85065081]
 [ 0.3249197   1.37638192]]


# Questão 2

In [3]:
m1 = 73  # kg
k1 = 2600  # N/m
da = 0.002  # m
F0 = 13  # N
wop = 3*2*np.pi  # rad/s
m2, k2 = sp.symbols("m2 k2", real=True, positive=True)  # A descobrir

Temos que 

$$
[M]\{\ddot{X}\} + [C]\{\ddot{X}\} + [K]\{X\} = \{F_0\} \cdot \exp(i\omega t)
$$
$$
\left(-\omega^2 [M] + i\omega [C] + [K]\right)\cdot \{X\} = \{F_0\}\exp \left(i\omega t\right)
$$
$$
\{X\} = \left(-\omega^2 [M] + i\omega [C] + [K]\right)^{-1} \cdot \{F_0\} \cdot \exp\left(i\omega t \right)
$$

Seja 

$$
[M] = \begin{bmatrix} m_1 & 0 \\ 0 & m_2
\end{bmatrix}
$$
$$
[C] = \begin{bmatrix} 0 & 0 \\ 0 & 0
\end{bmatrix}
$$
$$
[K] = \begin{bmatrix} k_1+k_2 & -k_2 \\ -k_2 & k_2
\end{bmatrix}
$$



In [4]:
K = sp.Matrix([[k1+k2, -k2],
               [-k2, k2]])
M = sp.Matrix([[m1, 0],
               [0, m2]])
print("K = ")
print(K)
print("M = ")
print(M)

K = 
Matrix([[k2 + 2600, -k2], [-k2, k2]])
M = 
Matrix([[73, 0], [0, m2]])


Então temos

$$
[K]-\omega^2 [M] = \begin{bmatrix}
k_1+k_2 -\omega^2 m_1 & -k_2 \\
-k_2 & k_2 -\omega^2 m_2
\end{bmatrix}
$$
$$
\det \left([K]-\omega^2 [M]\right) = m_1m_2 \omega^4 - \omega^2 \left(k_1m_2+k_2m_1+k_2m_2\right)+k_1k_2
$$
$$
\left([K]-\omega^2 [M]\right)^{-1} = \dfrac{1}{\det \left([K]-\omega^2[M]\right)} \cdot \begin{bmatrix}
k_2 -\omega^2 m_2 & k_2 \\ k_2 & k_1 + k_2 - \omega^2 m_1\end{bmatrix}
$$

In [5]:
Mat = sp.Matrix(K - wop**2 * M)
print("Mat = ")
print(Mat)
print("det(Mat) = ")
detMat = Mat.det()
print(detMat)
invMat = Mat.inv()
print("inv(Mat) = ")
invMat

Mat = 
Matrix([[k2 - 23337.3203660628, -k2], [-k2, k2 - 355.305758439217*m2]])
det(Mat) = 
-355.305758439217*k2*m2 - 23337.3203660628*k2 + 8291884.31260294*m2
inv(Mat) = 


Matrix([
[(-1.0*k2 + 355.305758439217*m2)/(355.305758439217*k2*m2 + 23337.3203660628*k2 - 8291884.31260294*m2),                     -1.0*k2/(355.305758439217*k2*m2 + 23337.3203660628*k2 - 8291884.31260294*m2)],
[                        -1.0*k2/(355.305758439217*k2*m2 + 23337.3203660628*k2 - 8291884.31260294*m2), (23337.3203660628 - 1.0*k2)/(355.305758439217*k2*m2 + 23337.3203660628*k2 - 8291884.31260294*m2)]])

Desta forma, teremos
$$
\begin{bmatrix}
X_1 \\ X_2
\end{bmatrix}
= \left([K]-\omega^2[M]\right)^{-1} \cdot \begin{bmatrix}
F_0 \\ 0 \end{bmatrix} \cdot \exp \left(i\omega t\right)
$$

De forma que temos

$$
X_1 = \dfrac{k_2 - \omega^2 m_2}{\det \left([K]-\omega^2[M]\right)} \cdot F_0 \cdot \exp \left(i\omega t\right)
$$
$$
X_2 = \dfrac{k_2}{\det \left([K]-\omega^2[M]\right)} \cdot F_0 \cdot \exp \left(i\omega t\right)
$$

In [6]:
F = sp.Matrix([[F0], [0]])
X1, X2 = invMat @ F
print("X1 = ")
print(X1)
print("X2 = ")
print(X2)

X1 = 
13*(-1.0*k2 + 355.305758439217*m2)/(355.305758439217*k2*m2 + 23337.3203660628*k2 - 8291884.31260294*m2)
X2 = 
-13.0*k2/(355.305758439217*k2*m2 + 23337.3203660628*k2 - 8291884.31260294*m2)


Tomando a frequência $\omega$ aquela que queremos anular o termo $X_1$ teremos

$$
k_2 - \omega^2 m_2 = 0 \Rightarrow k_2 = m_2\omega^2 \ \ \ \ \ \ \ \ \ (2)
$$

In [7]:
K = K.subs(k2, m2*wop**2)
M = M.subs(k2, m2*wop**2)
Mat = K - wop**2 * M
invMat = Mat.inv()
_, X2 = sp.Abs(invMat @ F)
X2

0.0365882052041775/m2

Agora, teremos a limitação devido ao deslocamento da massa 2. Isto se traduz como

$$
\dfrac{k_2}{\det\left([K]-\omega^2[M]\right)}=\dfrac{X_2}{F_0} \le \dfrac{\Delta a}{F_0}\ \ \ \ \ \ \ \ (3)
$$

Então resolvendo utilizando o ```sympy```

In [8]:
solution = sp.solve(X2-da, m2)
m2 = solution[0]
k2 = wop**2 * m2
print("m2 = %.3f" % m2)
print("k2 = %.3f" % k2)

m2 = 18.294
k2 = 6500.000


Logo, temos os valores da mola e da massa

$$
k_2 = 6500 N/m
$$
$$
m_2 = 18.294 kg
$$